src/main/java/cat/Parser.java

```diff
@@ -0,0 +1,221 @@
+        boolean isRange = value.matches("\\d+-\\d+");
+        if (isRange) {
+            String[] rangeEnds = value.split("-");
+            for (int i = Integer.parseInt(rangeEnds[0]) - 1; i <= Integer.parseInt(rangeEnds[1]) - 1; i++) {
+                indices.add(i);
+            }
+        } else {
+            indices.add(Integer.parseInt(value) - 1);
+        }
+    }
+
+    private static Components parseComponents(String data) throws InvalidCommandData {
+        assert data != null : "data must not be null";
+
+        HashMap<String, StringBuilder> parameterBuilders = new HashMap<>();
+
+        String key = DESCRIPTION_IDENTIFIER;
+        String[] words = data.split(" +");
+        for (String word : words) {
+            if (word.startsWith("/")) {
+                validateBuilderNotNull(parameterBuilders, key);
+
+                key = word;
+                continue;
+            }
+            parameterBuilders
+                    .compute(key, (k, v) -> (v == null) ? new StringBuilder(word) : v.append(" ").append(word));
+        }
+
+        HashMap<String, String> parameters = new HashMap<>();
+        parameterBuilders.forEach((k, v) -> parameters.put(k, v.toString()));
+
+        String description = parameters.remove(DESCRIPTION_IDENTIFIER);
+        assert description != null : "Description must not be null after extracting from parsed parameters";
+
+        return new Components(description, parameters);
+    }
+
+    private static void validateBuilderNotNull(
+            HashMap<String, StringBuilder> componentBuilders, String key) throws InvalidCommandData {
+        boolean isPreviousBuilderNull = componentBuilders.get(key) == null;
+        if (isPreviousBuilderNull) {
+            throw new InvalidCommandData(key);
+        }
+    }
+
+    /**
+     * Exception when the command is unknown.
+     */
+    public static class InvalidCommandType extends Exception {
+        private final String command;
+
+        /**
+         * Creates an exception about an unknown command string.
+         *
+         * @param command The name of the unknown command
+         */
+        public InvalidCommandType(String command) {
+            super("Command \"" + command + "\" is invalid or not yet implemented.");
+            this.command = command;
+        }
+
+        public String getCommand() {
+            return command;
+        }
+    }
+
+    /**
+     * Exception when the data given to a particular
+     * command doesn't match what is expected. The problem
+     * can either be that no parameters are given, or that
+     * a parameter is missing.
+     */
+    public static class InvalidCommandData extends Exception {
+        public InvalidCommandData() {
+            super("Parameters to command not given.");
+        }
+
+        public InvalidCommandData(String key) {
+            super("No value given to " + key);
+        }
+    }
```

Comment:
Hmm I think that makes sense. I was making the exceptions near where they are thrown from, but it might be better to centralise them. Well I'm still not 100% sold on that, but I think having a CatException to indicate that this is for internal errors is something I'll absolutely implement. Thanks!